platform/DeepSleepLock.h

```diff
@@ -36,29 +38,47 @@ namespace mbed {
   * @endcode
   */
 class DeepSleepLock {
+private:
+    uint16_t _lock_count;
+
 public:
-    DeepSleepLock()
+    DeepSleepLock(): _lock_count(1)
     {
         sleep_manager_lock_deep_sleep();
     }
 
     ~DeepSleepLock()
     {
-        sleep_manager_unlock_deep_sleep();
+        if (_lock_count) {
+            sleep_manager_unlock_deep_sleep();
+        }
     }
 
     /** Mark the start of a locked deep sleep section
      */
     void lock()
     {
-        sleep_manager_lock_deep_sleep();
+        uint16_t count = core_util_atomic_incr_u16(&_lock_count, 1);
+        if (1 == count) {
+            sleep_manager_lock_deep_sleep();
+        }
+        if (0 == count) {
+            error("DeepSleepLock underflow (< 0)");
```

Comment:
Is this error report correct? if it is 0 , it already wrapped around (we increment here) ? The conditions seems like reversed (here and in unlock) ?

Reply:
It should be. The function `core_util_atomic_incr_u16` returns the new value, so if lock (which should only make the value bigger) sets the value to 0 then an overflow has occurred.

Reply:
OK, what about both being underflow? in the case of 0 here, it is overflow